api/include/opentelemetry/trace/trace_state.h

```diff
@@ -45,51 +45,154 @@ class TraceState
 
   // Returns false if no such key, otherwise returns true and populates value.
-  bool Get(nostd::string_view key, nostd::string_view value) const noexcept { return false; }
+  bool Get(nostd::string_view key, nostd::string_view value) noexcept
+  {
+    for (int i = 0; i < num_entries_; i++)
+    {
+      if (strcmp(key.data(), entries_[i].GetKey().data()) == 0)
+      {
+        entries_[i].SetValue(value);
+        return true;
+      }
+    }
+    return false;
+  }
 
   // Creates an Entry for the key-value pair and adds it to entries.
   // If value is null, this function is a no-op.
-  void Set(nostd::string_view key, nostd::string_view value) const noexcept;
+  void Set(nostd::string_view key, nostd::string_view value) noexcept
+  {
+    if (value.data() == NULL)
+      return;
+
+    Entry entry(key, value);
+
+    if (num_entries_ < kMaxKeyValuePairs)
+    {
+      entries_[num_entries_] = entry;
+      num_entries_++;
+    }
+  }
 
   // Returns true if there are no keys, false otherwise.
-  bool Empty() const noexcept { return true; }
+  bool Empty() const noexcept { return num_entries_ == 0; }
 
   // Returns a span of all the entries. The TraceState object must outlive the span.
-  nostd::span<Entry> Entries() const noexcept { return {}; }
+  nostd::span<Entry> Entries() noexcept { return nostd::span<Entry>(entries_.get(), num_entries_); }
 
   // Returns whether key is a valid key. See https://www.w3.org/TR/trace-context/#key
-  static bool IsValidKey(nostd::string_view key);
+  static bool IsValidKey(nostd::string_view key)
+  {
+    if (key.empty() || key.size() > kKeyMaxSize || !IsLowerCaseAlphaOrDigit(key[0]))
+    {
+      return false;
+    }
+
+    int ats     = 0;
+    const int n = key.size();
+
+    for (int i = 0; i < n; ++i)
+    {
+      char c = key[i];
+      if (!IsLowerCaseAlphaOrDigit(c) && c != '_' && c != '-' && c != '@' && c != '*' && c != '/')
+      {
+        return false;
+      }
+      if ((c == '@') && (++ats > 1))
+      {
+        return false;
+      }
+    }
+    return true;
+  }
 
   // Returns whether value is a valid value. See https://www.w3.org/TR/trace-context/#value
-  static bool IsValidValue(nostd::string_view value);
+  static bool IsValidValue(nostd::string_view value)
+  {
+    if (value.empty() || value.size() > kValueMaxSize)
+    {
+      return false;
+    }
+
+    const int n = value.size();
+
+    for (int i = 0; i < n; ++i)
+    {
+      char c = value[i];
+      if (c < ' ' || c > '~' || c == ',' || c == '=')
+      {
+        return false;
+      }
+    }
+    return true;
+  }
 
 private:
   // Store entries in a C-style array to avoid using std::array or std::vector.
-  Entry entries_[kMaxKeyValuePairs];
+  std::unique_ptr<Entry[]> entries_;
 
   // Maintain the number of entries in entries_. Must be in the range [0, kMaxKeyValuePairs].
   int num_entries_;
+
+  static bool IsLowerCaseAlphaOrDigit(char c)
+  {
+    return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
```

Comment:
You could also use `return isdigit(c) || islower(c);` here. See http://www.cplusplus.com/reference/cctype/.